backend/app/bridge/host.py

```
@@ -267,6 +268,269 @@ def _update_token_expiry(token) -> None:
+def _id_has_legacy_fragment(value: object) -> bool:
+    """Whether ``$id`` carries a NON-EMPTY fragment — draft-07's plain-name anchor form
+    (``"https://example.test/tool#thing"``), which 2020-12 does not allow: there ``$id`` may
+    have no fragment or an empty one, and the anchor role moved to ``$anchor``. Relabeling
+    such a root would leave behind an ``$id`` the 2020-12 meta-schema rejects, so the strict
+    client goes on refusing the tool — swapping its complaint rather than resolving it
+    (#124 review). A non-string ``$id`` is malformed rather than legacy, and is left to the
+    same pass-through the malformed-``$schema`` case gets."""
+    return isinstance(value, str) and "#" in value and not value.endswith("#")
+
+
+def _has_incompatible_draft07_construct(schema: object, *, is_root: bool = True) -> bool:
```

Ruff (0.16.1)

[warning] Too many return statements (12 > 6)

(PLR0911)